File: clojure/beer-song/beer.clj
(ns beer
  (:use [clojure.string :only [capitalize]]))

(defn- bottles [num]
  (cond
   (zero? num) "no more bottles of beer"
   (= 1 num)   "1 bottle of beer"
   :else       (str num " bottles of beer")))

(defn- on-the-wall [num]
  (let [phrase (bottles num)]
    (str (capitalize phrase) " on the wall, " phrase ".")))

(defn- take-down [num]
  (if (>= num 0)
    (str "Take "
         (if (zero? num) "it" "one")
         " down and pass it around, "
         (bottles num)
         " on the wall.")
    "Go to the store and buy some more, 99 bottles of beer on the wall."))

(defn verse
  "Sing a verse of 99 bottles of beer on the wall"
  [num]
  (str (on-the-wall num) "\n" (take-down (dec num)) "\n"))

(defn sing
  "Sing verses from start down to end (inclusive)"
  ([start] (sing start 0))
  ([start end] (apply str (interleave
                           (map verse (range start (dec end) -1))
                           (repeat "\n")))))

File: elixir/beer-song/beer.exs
defmodule Beer do
  defp bottles(0), do: "no more bottles of beer"
  defp bottles(1), do: "1 bottle of beer"
  defp bottles(n), do: <<integer_to_binary(n) :: binary, " bottles of beer">>

  defp on_the_wall(n) do
    phrase = bottles(n)
    <<String.capitalize(phrase) :: binary,
      " on the wall, ",
      phrase :: binary,
      ".">>
  end

  defp take_down(n) when n >= 0 do
    <<"Take ",
      (if (n == 0) do "it" else "one" end) :: binary,
      " down and pass it around, ",
      bottles(n) :: binary,
      " on the wall.">>
  end
  defp take_down(_) do
    "Go to the store and buy some more, 99 bottles of beer on the wall."
  end

  @doc "Sing the verse of 99 bottles of beer on the wall given n bottles"
  def verse(n) do
    <<on_the_wall(n) :: binary,
      "\n",
      take_down(n - 1) :: binary,
      "\n">>
  end

  @doc "Sing verses from start_verse down to end_verse (inclusive)"
  def sing(start_verse, end_verse // 0) do
    verses = Enum.to_list(start_verse..end_verse)
    bc n inlist verses, do: <<verse(n) :: binary, "\n">>
  end
end